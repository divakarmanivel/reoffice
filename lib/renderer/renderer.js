"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WordRenderer = void 0;
var _reactReconciler = _interopRequireDefault(require("react-reconciler"));
var _createElement = require("../utils/createElement");
var _RendererHostConfig;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return _typeof(key) === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (_typeof(input) !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (_typeof(res) !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); } /* eslint-disable */
var RendererHostConfig = (_RendererHostConfig = {
  appendInitialChild: function appendInitialChild(parentInstance, child) {
    if (parentInstance.appendChild) {
      parentInstance.appendChild(child);
    } else {
      parentInstance.document = child;
    }
  },
  createInstance: function createInstance(type, props, internalInstanceHandle) {
    return (0, _createElement.createElement)(type, props);
  },
  createTextInstance: function createTextInstance(text, rootContainerInstance, internalInstanceHandle) {
    return text;
  },
  finalizeInitialChildren: function finalizeInitialChildren(wordElement, type, props) {
    return false;
  },
  getPublicInstance: function getPublicInstance(inst) {
    return inst;
  },
  prepareForCommit: function prepareForCommit() {
    // noop
  },
  prepareUpdate: function prepareUpdate(wordElement, type, oldProps, newProps) {
    return true;
  },
  resetAfterCommit: function resetAfterCommit() {
    // noop
  },
  resetTextContent: function resetTextContent(wordElement) {
    // Reoffice does not have a text node like DOM
  }
}, _defineProperty(_RendererHostConfig, "createInstance", function createInstance(type, props, internalInstanceHandle) {
  // 'internalInstanceHandle' is not transparent here. So use host context methods
  // to get data from roots
  return (0, _createElement.createElement)(type, props);
}), _defineProperty(_RendererHostConfig, "getRootHostContext", function getRootHostContext(instance) {
  // getHostContextNode here updates the internal state of createElement and stores a ref to current instance
  return (0, _createElement.getHostContextNode)(instance);
}), _defineProperty(_RendererHostConfig, "getChildHostContext", function getChildHostContext() {
  var context = {};
  return context;
}), _defineProperty(_RendererHostConfig, "shouldSetTextContent", function shouldSetTextContent(type, props) {
  return false; // Reoffice does not have a text node like DOM
}), _defineProperty(_RendererHostConfig, "now", function now() {
  return Date.now;
}), _defineProperty(_RendererHostConfig, "supportsMutation", true), _defineProperty(_RendererHostConfig, "appendChild", function appendChild(parentInstance, child) {
  if (parentInstance.appendChild) {
    parentInstance.appendChild(child);
  } else {
    parentInstance.document = child;
  }
}), _defineProperty(_RendererHostConfig, "appendChildToContainer", function appendChildToContainer(parentInstance, child) {
  if (parentInstance.appendChild) {
    parentInstance.appendChild(child);
  } else {
    parentInstance.document = child;
  }
}), _defineProperty(_RendererHostConfig, "removeChild", function removeChild(parentInstance, child) {
  parentInstance.removeChild(child);
}), _defineProperty(_RendererHostConfig, "removeChildFromContainer", function removeChildFromContainer(parentInstance, child) {
  parentInstance.removeChild(child);
}), _defineProperty(_RendererHostConfig, "insertBefore", function insertBefore(parentInstance, child, beforeChild) {
  // noop
}), _defineProperty(_RendererHostConfig, "commitUpdate", function commitUpdate(instance, updatePayload, type, oldProps, newProps) {
  // noop
}), _defineProperty(_RendererHostConfig, "commitMount", function commitMount(instance, updatePayload, type, oldProps, newProps) {
  // noop
}), _defineProperty(_RendererHostConfig, "commitTextUpdate", function commitTextUpdate(textInstance, oldText, newText) {
  textInstance.children = newText;
}), _defineProperty(_RendererHostConfig, "clearContainer", function clearContainer(container) {
  // noop
}), _RendererHostConfig);
var WordRenderer = (0, _reactReconciler["default"])(RendererHostConfig);
exports.WordRenderer = WordRenderer;