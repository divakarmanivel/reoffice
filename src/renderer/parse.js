/**
 * Render the component (passed to docx generator instance)
 * This is similar to react-pdf https://github.com/diegomura/react-pdf/blob/master/packages/react-pdf/src/index.js
 * @param {Object} input Root component
 */
const parse = (input) => {
  async function parseComponent(inputComponent) {
    // This property is accessed due to https://github.com/divakarmanivel/reoffice/blob/master/src/renderer/renderer.js#L32
    const { document } = inputComponent;
    await document.render();
    // Return the input component again because we rendered the children
    // which weren't wrapped inside a parent.
    // We async called the render method on all of the children.
    return inputComponent;
  }

  async function toBuffer() {
    return await parseComponent(input);
  }

  return {
    toBuffer,
  };
};

export default parse;
