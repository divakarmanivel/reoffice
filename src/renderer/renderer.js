/* eslint-disable */

import ReactReconciler from 'react-reconciler';
import { createElement, getHostContextNode } from '../utils/createElement';

const RendererHostConfig = {
	appendInitialChild(parentInstance, child) {
		if (parentInstance.appendChild) {
			parentInstance.appendChild(child);
		} else {
			parentInstance.document = child;
		}
	},

	createInstance(type, props, internalInstanceHandle) {
		return createElement(type, props);
	},

	createTextInstance(text, rootContainerInstance, internalInstanceHandle) {
		return text;
	},

	finalizeInitialChildren(wordElement, type, props) {
		return false;
	},

	getPublicInstance(inst) {
		return inst;
	},

	prepareForCommit() {
		// noop
	},

	prepareUpdate(wordElement, type, oldProps, newProps) {
		return true;
	},

	resetAfterCommit() {
		// noop
	},

	resetTextContent(wordElement) {
		// Reoffice does not have a text node like DOM
	},

	// If you've such use case where you need to provide data from the root instance,
	// see the below example. This may help you in creating your own custom renderer
	
	createInstance(type, props, internalInstanceHandle) {
		// 'internalInstanceHandle' is not transparent here. So use host context methods
		// to get data from roots
		return createElement(type, props);
	},

	// Use this current instance to pass data from root
	getRootHostContext(instance) {
		// getHostContextNode here updates the internal state of createElement and stores a ref to current instance
    	return getHostContextNode(instance);
	},

	getChildHostContext() {
		let context = {};
		return context;
	},

	shouldSetTextContent(type, props) {
		return false; // Reoffice does not have a text node like DOM
	},

	now: () => Date.now,

	supportsMutation: true,

	appendChild(parentInstance, child) {
		if (parentInstance.appendChild) {
			parentInstance.appendChild(child);
		} else {
			parentInstance.document = child;
		}
	},

	appendChildToContainer(parentInstance, child) {
		if (parentInstance.appendChild) {
			parentInstance.appendChild(child);
		} else {
			parentInstance.document = child;
		}
	},

	removeChild(parentInstance, child) {
		parentInstance.removeChild(child);
	},

	removeChildFromContainer(parentInstance, child) {
		parentInstance.removeChild(child);
	},

	insertBefore(parentInstance, child, beforeChild) {
		// noop
	},

	commitUpdate(instance, updatePayload, type, oldProps, newProps) {
		// noop
	},

	commitMount(instance, updatePayload, type, oldProps, newProps) {
		// noop
	},

	commitTextUpdate(textInstance, oldText, newText) {
		textInstance.children = newText;
	},

	clearContainer(container) {
		// noop
	},
}

const WordRenderer = ReactReconciler(RendererHostConfig);

export { WordRenderer }