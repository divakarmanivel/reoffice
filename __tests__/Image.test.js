import React from 'react';
import { Image } from '../src';
import { testRenderer as render } from '../src/renderer/render';

it('sanity check', () => {
  expect(render(<Image />)).toMatchSnapshot();
});

it('should render an image', () => {
  const App = () => (
    <Image src='../src/concr.png' />
  );

  expect(render(<App />)).toMatchSnapshot();
});

it('should render an image and align it', () => {
  const App = () => (
    <Image src='../src/concr.png' align='justify' />
  );

  expect(render(<App />)).toMatchSnapshot();
});

it('should render an image and adjust it size with style prop', () => {
  const App = () => (
    <Image src='../src/concr.png' style={{ width: '400', height: '400' }} />
  );

  expect(render(<App />)).toMatchSnapshot();
});

it('should render an image and adjust it size with width and height prop', () => {
  const App = () => (
    <Image src='../src/concr.png' width='400' height='400' />
  );

  expect(render(<App />)).toMatchSnapshot();
});
