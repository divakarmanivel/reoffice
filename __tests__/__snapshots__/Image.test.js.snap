// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sanity check 1`] = `
WordDocument {
  "doc": officegen {
    "_events": {
      "afterGen": [Function],
      "beforeGen": [Function],
      "clearData": [Function],
      "clearDoc": [Function],
      "clearDocType": [Function],
    },
    "_eventsCount": 5,
    "_maxListeners": undefined,
    "addPageBreak": [Function],
    "addResourceToParse": [Function],
    "cbMakeDocxDocument": [Function],
    "createByJson": [Function],
    "createJson": [Function],
    "createListOfDots": [Function],
    "createListOfNumbers": [Function],
    "createNestedOrderedList": [Function],
    "createNestedUnOrderedList": [Function],
    "createP": [Function],
    "createTable": [Function],
    "data": [],
    "generate": [Function],
    "getFooter": [Function],
    "getHeader": [Function],
    "info": {
      "cp:category": {
        "data": "",
        "def_data": "",
        "element": "cp:category",
      },
      "cp:contentStatus": {
        "data": "",
        "def_data": "",
        "element": "cp:contentStatus",
      },
      "cp:keywords": {
        "data": "",
        "def_data": "",
        "element": "cp:keywords",
      },
      "dc:description": {
        "data": "",
        "def_data": "",
        "element": "dc:description",
      },
      "dc:subject": {
        "data": "",
        "def_data": "",
        "element": "dc:subject",
      },
      "dc:title": {
        "data": "",
        "def_data": "",
        "element": "dc:title",
      },
    },
    "options": {
      "type": "docx",
    },
    "putPageBreak": [Function],
    "setDescription": [Function],
    "setDocCategory": [Function],
    "setDocKeywords": [Function],
    "setDocStatus": [Function],
    "setDocSubject": [Function],
    "setDocTitle": [Function],
    "setTheme": [Function],
    "startNewDoc": [Function],
    "theme": null,
    Symbol(kCapture): false,
  },
  "document": Image {
    "children": [],
    "name": "Image",
    "parent": null,
    "props": {},
    "root": [Circular],
  },
}
`;

exports[`should render an image 1`] = `
WordDocument {
  "doc": officegen {
    "_events": {
      "afterGen": [Function],
      "beforeGen": [Function],
      "clearData": [Function],
      "clearDoc": [Function],
      "clearDocType": [Function],
    },
    "_eventsCount": 5,
    "_maxListeners": undefined,
    "addPageBreak": [Function],
    "addResourceToParse": [Function],
    "cbMakeDocxDocument": [Function],
    "createByJson": [Function],
    "createJson": [Function],
    "createListOfDots": [Function],
    "createListOfNumbers": [Function],
    "createNestedOrderedList": [Function],
    "createNestedUnOrderedList": [Function],
    "createP": [Function],
    "createTable": [Function],
    "data": [],
    "generate": [Function],
    "getFooter": [Function],
    "getHeader": [Function],
    "info": {
      "cp:category": {
        "data": "",
        "def_data": "",
        "element": "cp:category",
      },
      "cp:contentStatus": {
        "data": "",
        "def_data": "",
        "element": "cp:contentStatus",
      },
      "cp:keywords": {
        "data": "",
        "def_data": "",
        "element": "cp:keywords",
      },
      "dc:description": {
        "data": "",
        "def_data": "",
        "element": "dc:description",
      },
      "dc:subject": {
        "data": "",
        "def_data": "",
        "element": "dc:subject",
      },
      "dc:title": {
        "data": "",
        "def_data": "",
        "element": "dc:title",
      },
    },
    "options": {
      "type": "docx",
    },
    "putPageBreak": [Function],
    "setDescription": [Function],
    "setDocCategory": [Function],
    "setDocKeywords": [Function],
    "setDocStatus": [Function],
    "setDocSubject": [Function],
    "setDocTitle": [Function],
    "setTheme": [Function],
    "startNewDoc": [Function],
    "theme": null,
    Symbol(kCapture): false,
  },
  "document": Image {
    "children": [],
    "name": "Image",
    "parent": null,
    "props": {
      "src": "../src/concr.png",
    },
    "root": [Circular],
  },
}
`;

exports[`should render an image and adjust it size with style prop 1`] = `
WordDocument {
  "doc": officegen {
    "_events": {
      "afterGen": [Function],
      "beforeGen": [Function],
      "clearData": [Function],
      "clearDoc": [Function],
      "clearDocType": [Function],
    },
    "_eventsCount": 5,
    "_maxListeners": undefined,
    "addPageBreak": [Function],
    "addResourceToParse": [Function],
    "cbMakeDocxDocument": [Function],
    "createByJson": [Function],
    "createJson": [Function],
    "createListOfDots": [Function],
    "createListOfNumbers": [Function],
    "createNestedOrderedList": [Function],
    "createNestedUnOrderedList": [Function],
    "createP": [Function],
    "createTable": [Function],
    "data": [],
    "generate": [Function],
    "getFooter": [Function],
    "getHeader": [Function],
    "info": {
      "cp:category": {
        "data": "",
        "def_data": "",
        "element": "cp:category",
      },
      "cp:contentStatus": {
        "data": "",
        "def_data": "",
        "element": "cp:contentStatus",
      },
      "cp:keywords": {
        "data": "",
        "def_data": "",
        "element": "cp:keywords",
      },
      "dc:description": {
        "data": "",
        "def_data": "",
        "element": "dc:description",
      },
      "dc:subject": {
        "data": "",
        "def_data": "",
        "element": "dc:subject",
      },
      "dc:title": {
        "data": "",
        "def_data": "",
        "element": "dc:title",
      },
    },
    "options": {
      "type": "docx",
    },
    "putPageBreak": [Function],
    "setDescription": [Function],
    "setDocCategory": [Function],
    "setDocKeywords": [Function],
    "setDocStatus": [Function],
    "setDocSubject": [Function],
    "setDocTitle": [Function],
    "setTheme": [Function],
    "startNewDoc": [Function],
    "theme": null,
    Symbol(kCapture): false,
  },
  "document": Image {
    "children": [],
    "name": "Image",
    "parent": null,
    "props": {
      "src": "../src/concr.png",
      "style": {
        "height": "400",
        "width": "400",
      },
    },
    "root": [Circular],
  },
}
`;

exports[`should render an image and adjust it size with width and height prop 1`] = `
WordDocument {
  "doc": officegen {
    "_events": {
      "afterGen": [Function],
      "beforeGen": [Function],
      "clearData": [Function],
      "clearDoc": [Function],
      "clearDocType": [Function],
    },
    "_eventsCount": 5,
    "_maxListeners": undefined,
    "addPageBreak": [Function],
    "addResourceToParse": [Function],
    "cbMakeDocxDocument": [Function],
    "createByJson": [Function],
    "createJson": [Function],
    "createListOfDots": [Function],
    "createListOfNumbers": [Function],
    "createNestedOrderedList": [Function],
    "createNestedUnOrderedList": [Function],
    "createP": [Function],
    "createTable": [Function],
    "data": [],
    "generate": [Function],
    "getFooter": [Function],
    "getHeader": [Function],
    "info": {
      "cp:category": {
        "data": "",
        "def_data": "",
        "element": "cp:category",
      },
      "cp:contentStatus": {
        "data": "",
        "def_data": "",
        "element": "cp:contentStatus",
      },
      "cp:keywords": {
        "data": "",
        "def_data": "",
        "element": "cp:keywords",
      },
      "dc:description": {
        "data": "",
        "def_data": "",
        "element": "dc:description",
      },
      "dc:subject": {
        "data": "",
        "def_data": "",
        "element": "dc:subject",
      },
      "dc:title": {
        "data": "",
        "def_data": "",
        "element": "dc:title",
      },
    },
    "options": {
      "type": "docx",
    },
    "putPageBreak": [Function],
    "setDescription": [Function],
    "setDocCategory": [Function],
    "setDocKeywords": [Function],
    "setDocStatus": [Function],
    "setDocSubject": [Function],
    "setDocTitle": [Function],
    "setTheme": [Function],
    "startNewDoc": [Function],
    "theme": null,
    Symbol(kCapture): false,
  },
  "document": Image {
    "children": [],
    "name": "Image",
    "parent": null,
    "props": {
      "height": "400",
      "src": "../src/concr.png",
      "width": "400",
    },
    "root": [Circular],
  },
}
`;

exports[`should render an image and align it 1`] = `
WordDocument {
  "doc": officegen {
    "_events": {
      "afterGen": [Function],
      "beforeGen": [Function],
      "clearData": [Function],
      "clearDoc": [Function],
      "clearDocType": [Function],
    },
    "_eventsCount": 5,
    "_maxListeners": undefined,
    "addPageBreak": [Function],
    "addResourceToParse": [Function],
    "cbMakeDocxDocument": [Function],
    "createByJson": [Function],
    "createJson": [Function],
    "createListOfDots": [Function],
    "createListOfNumbers": [Function],
    "createNestedOrderedList": [Function],
    "createNestedUnOrderedList": [Function],
    "createP": [Function],
    "createTable": [Function],
    "data": [],
    "generate": [Function],
    "getFooter": [Function],
    "getHeader": [Function],
    "info": {
      "cp:category": {
        "data": "",
        "def_data": "",
        "element": "cp:category",
      },
      "cp:contentStatus": {
        "data": "",
        "def_data": "",
        "element": "cp:contentStatus",
      },
      "cp:keywords": {
        "data": "",
        "def_data": "",
        "element": "cp:keywords",
      },
      "dc:description": {
        "data": "",
        "def_data": "",
        "element": "dc:description",
      },
      "dc:subject": {
        "data": "",
        "def_data": "",
        "element": "dc:subject",
      },
      "dc:title": {
        "data": "",
        "def_data": "",
        "element": "dc:title",
      },
    },
    "options": {
      "type": "docx",
    },
    "putPageBreak": [Function],
    "setDescription": [Function],
    "setDocCategory": [Function],
    "setDocKeywords": [Function],
    "setDocStatus": [Function],
    "setDocSubject": [Function],
    "setDocTitle": [Function],
    "setTheme": [Function],
    "startNewDoc": [Function],
    "theme": null,
    Symbol(kCapture): false,
  },
  "document": Image {
    "children": [],
    "name": "Image",
    "parent": null,
    "props": {
      "align": "justify",
      "src": "../src/concr.png",
    },
    "root": [Circular],
  },
}
`;
